runner/src/diffs.rs

```diff
@@ -0,0 +1,298 @@
+    if let Ok(file) = std::fs::File::open(proto_file_path) {
+        let lines = BufReader::new(file).lines();
+        let re = regex::Regex::new(r#"import "(.*)";"#).unwrap();
+
+        // Scan the lines in the file line by line to find all the imports.
+        for line in lines {
+            let line = line.expect("could not read line");
+            if let Some(imported) = re.captures(&line).map(|c| c[1].to_string()) {
+                imported_protos.push(imported);
+            }
+        }
+    }
+
+    imported_protos
+}
+
+/// Path to the `Cargo.toml` files for all crates that are either directly modified or have a
+/// dependency to a modified crate.
+pub fn all_affected_crates(commits: &Commits) -> ModifiedContent {
+    match commits.commits {
+        None => ModifiedContent {
+            files: vec![],
+            all_modified: true,
+        },
+        _ => {
+            let crate_manifest_files = crate_manifest_files();
+            let mut affected_crates =
+                HashSet::<String>::from_iter(directly_modified_crates(commits).into_iter());
+
+            let crates_affected_by_protos = crates_affected_by_protos(&affected_protos(commits));
+            affected_crates = affected_crates
+                .union(&crates_affected_by_protos)
+                .cloned()
+                .collect();
+
+            for crate_path in crate_manifest_files {
+                add_affected_crates(crate_path, &mut affected_crates);
+            }
+
+            ModifiedContent {
+                files: affected_crates.iter().cloned().collect(),
+                all_modified: false,
+            }
+        }
+    }
+}
+
+/// Returns the paths to `Cargo.toml` files of crates affected by the changed proto files.
+fn crates_affected_by_protos(affected_protos: &[String]) -> HashSet<String> {
+    source_files()
+        .filter(|path| to_string(path.clone()).ends_with("build.rs"))
+        .filter(|path| {
+            for proto in affected_protos {
+                if file_contains(path, proto) {
+                    return true;
+                }
+            }
+            false
+        })
+        .map(to_string)
+        .map(|build_path| find_crate_toml_file(build_path).unwrap())
+        .collect()
+}
+
+// Checks if `crate_toml_path` has a direct or indirect dependency to any of the crates in
+// `affected_crates_toml_path`. If so, adds `crate_toml_path` and any of its affected dependencies
+// to `affected_crates_toml_path`.
+fn add_affected_crates(crate_toml_path: PathBuf, affected_crates_toml_path: &mut HashSet<String>) {
+    if affected_crates_toml_path.contains(&to_string(crate_toml_path.clone())) {
+        return;
+    }
+    let deps = get_local_dependencies(&crate_toml_path);
+    for dep in deps {
+        if !affected_crates_toml_path.contains(&dep) {
+            let dep_path = PathBuf::from(dep.clone());
+            add_affected_crates(dep_path.clone(), affected_crates_toml_path)
```

Comment:
Seeing that this checks all dependencies, including dev and build dependencies, it seems that it is possible that it could pick up circular dependencies if it does not keep track of all visited crates (not just affected crates).